Bump GitHub Actions versions and use Go 1.22 and 1.21

Lekensteyn opened the pull request:
* Use latest macOS and Windows versions instead of referring to an outdated version that might be removed later. See the table at https://docs.github.com/en/actions/using-github-hosted-runners/about-github-hosted-runners/about-github-hosted-runners#standard-github-hosted-runners-for-public-repositories
* Use the `stable` and `oldstable` aliases instead of hard-coding specific point releases for `actions/setup-go`. The arm64 builder is using a Docker image, so use the `1.22` image tag. We can use the `latest` tag in the future to avoid having to update the version.
* Bump `github/codeql-action` actions to v3 as v2 is deprecated: https://github.com/github/codeql-action?tab=readme-ov-file#supported-versions-of-the-codeql-action
* Bump Go version in go.mod to the minimum Go version supported by Go.
* Bump golangci-lint for Go 1.22 and fix some linting errors.

The following SA1019 errors were ignored and need to be fixed later:

    (3x) SA1019: elliptic.Marshal has been deprecated since Go 1.21:
        for ECDH, use the crypto/ecdh package.
        This function returns an encoding equivalent to that of PublicKey.Bytes in crypto/ecdh.  (staticcheck)

    (2x) SA1019: elliptic.GenerateKey has been deprecated since Go 1.21:
        for ECDH, use the GenerateKey methods of the [crypto/ecdh] package;
        for ECDSA, use the GenerateKey function of the crypto/ecdsa package.  (staticcheck)

    (3x) SA1019: elliptic.Unmarshal has been deprecated since Go 1.21:
        for ECDH, use the crypto/ecdh package.
        This function accepts an encoding equivalent to that of the NewPublicKey methods in crypto/ecdh.  (staticcheck)

    (1x) SA1019: params.ScalarBaseMult has been deprecated since Go 1.21:
        the [CurveParams] methods are deprecated and are not guaranteed to provide any security property.
        For ECDH, use the [crypto/ecdh] package.
        For ECDSA, use the [crypto/ecdsa] package with a [Curve] value returned directly from [P224], [P256], [P384], or [P521].  (staticcheck)

Lekensteyn commented on .etc/golangci.yml:
```diff
@@ -39,13 +39,28 @@ linters:
   nestif:
     min-complexity: 6
+  goconst:
+    ignore-tests: true
```
Addresses
```
::error file=group/group_test.go,line=27,col=11::string `/Mul` has 3 occurrences, make it a constant (goconst)
::error file=group/group_test.go,line=25,col=11::string `/Add` has 3 occurrences, make it a constant (goconst)
```

armfazh replied:
we can add constants as is suggested, instead of ignoring everything.

Lekensteyn commented on blindsign/blindrsa/brsa_test.go:
```diff
@@ -20,8 +20,8 @@ import (
 // 2048-bit RSA private key
-const testPrivateKey = `
------BEGIN RSA PRIVATE KEY-----
+var testPrivateKey = testingKey(`
```
Addresses
```
::error file=blindsign/blindrsa/brsa_test.go,line=23,col=7::G101: Potential hardcoded credentials: RSA private key (gosec)
```

Follows a convention from the standard library (using a `testingKey` function like this).

Lekensteyn commented on .etc/golangci.yml:
```diff
@@ -39,13 +39,28 @@ linters:
+        allow:
+          - $gostd
+          - github.com/bwesterb/go-ristretto
+          - github.com/cloudflare/circl
```
Addresses 327 "errors", 326 due to circl usage, one from ristretto:
```
::error file=xof/xof.go,line=12,col=2::import 'github.com/cloudflare/circl/internal/sha3' is not allowed from list 'Main' (depguard)
...
::error file=group/ristretto255.go,line=10,col=2::import 'github.com/bwesterb/go-ristretto' is not allowed from list 'Main' (depguard)
```

Lekensteyn commented on blindsign/blindrsa/internal/common/rsa.go:
```diff
@@ -107,7 +106,7 @@ func decrypt(random io.Reader, priv *keys.BigPrivateKey, c *big.Int) (m *big.Int
 			r, err = rand.Int(random, priv.Pk.N)
 			if err != nil {
-				return
+				return nil, err
```
The above two changes address:
```
::error file=blindsign/blindrsa/internal/common/rsa.go,line=92,col=3::naked return in func `decrypt` with 48 lines of code (nakedret)
::error file=blindsign/blindrsa/internal/common/rsa.go,line=110,col=5::naked return in func `decrypt` with 48 lines of code (nakedret)
```

`m` has not been initialized up to this point, and `err` has been initialized. There are no writes to `m` in this loop, so returning `nil` is alright.

Lekensteyn commented on .etc/golangci.yml:
```diff
@@ -39,13 +39,28 @@ linters:
+  staticcheck:
+    # TODO: replace deprecated elliptic.Marshal, elliptic.GenerateKey,
+    # elliptic.Unmarshal, params.ScalarBaseMult before re-enabling SA1019.
+    checks: ["*", "-SA1019"]
```
This whole `staticcheck` override needs to be removed once the necessary API changes are made. I believe @armfazh already started on making those, so I just ignored the linter warning for now.

A summary of errors can be found in the commit message. The original errors were:
```
::error file=kem/hybrid/ckem.go,line=106,col=9::SA1019: elliptic.Marshal has been deprecated since Go 1.21: for ECDH, use the crypto/ecdh package. This function returns an encoding equivalent to that of PublicKey.Bytes in crypto/ecdh.  (staticcheck)
::error file=kem/hybrid/ckem.go,line=125,col=20::SA1019: elliptic.GenerateKey has been deprecated since Go 1.21: for ECDH, use the GenerateKey methods of the [crypto/ecdh] package; for ECDSA, use the GenerateKey function of the crypto/ecdsa package.  (staticcheck)
::error file=kem/hybrid/ckem.go,line=196,col=10::SA1019: elliptic.Unmarshal has been deprecated since Go 1.21: for ECDH, use the crypto/ecdh package. This function accepts an encoding equivalent to that of the NewPublicKey methods in crypto/ecdh.  (staticcheck)
::error file=ecc/p384/point_test.go,line=18,col=24::SA1019: params.ScalarBaseMult has been deprecated since Go 1.21: the [CurveParams] methods are deprecated and are not guaranteed to provide any security property. For ECDH, use the [crypto/ecdh] package. For ECDSA, use the [crypto/ecdsa] package with a [Curve] value returned directly from [P224], [P256], [P384], or [P521].  (staticcheck)
::error file=group/short.go,line=231,col=9::SA1019: elliptic.Marshal has been deprecated since Go 1.21: for ECDH, use the crypto/ecdh package. This function returns an encoding equivalent to that of PublicKey.Bytes in crypto/ecdh.  (staticcheck)
::error file=group/short.go,line=257,col=11::SA1019: elliptic.Unmarshal has been deprecated since Go 1.21: for ECDH, use the crypto/ecdh package. This function accepts an encoding equivalent to that of the NewPublicKey methods in crypto/ecdh.  (staticcheck)
::error file=hpke/shortkem.go,line=79,col=19::SA1019: elliptic.GenerateKey has been deprecated since Go 1.21: for ECDH, use the GenerateKey methods of the [crypto/ecdh] package; for ECDSA, use the GenerateKey function of the crypto/ecdsa package.  (staticcheck)
::error file=hpke/shortkem.go,line=99,col=10::SA1019: elliptic.Unmarshal has been deprecated since Go 1.21: for ECDH, use the crypto/ecdh package. This function accepts an encoding equivalent to that of the NewPublicKey methods in crypto/ecdh.  (staticcheck)
::error file=hpke/shortkem.go,line=120,col=9::SA1019: elliptic.Marshal has been deprecated since Go 1.21: for ECDH, use the crypto/ecdh package. This function returns an encoding equivalent to that of PublicKey.Bytes in crypto/ecdh.  (staticcheck)
```

Lekensteyn commented:
Changes since last push:

- Updated golangci-lint from v1.56.0 to v1.56.1.
- Use 1.22/1.21 instead of stable/oldstable, updated commit message. Based on preference expressed by @armfazh in https://github.com/cloudflare/circl/pull/476#discussion_r1482461506
